### PR TITLE
Some cleanings in the views (dashboard, layout, wall)

### PR DESCRIPTION
**Pull request by mohierf**
And cleanings in the hosts/services synthesis computation:
- say that an host/service is a problem when HARD non OK state
- make this consistent on all views (dashboard, currently, popovers, problems)

Manage a configurable minimum business impact
- set defaut to 2
- badges in the header bar only for problems > minimum BI

This has been tested in a test environment. Perharps some issues will raise in some specific case  ... but I will fix rapidly 😉

### CODE REVIEW

File: module/datamanager.py [hunk 1]
```diff
@@ -1000,21 +1024,42 @@ def get_overall_it_state(self, user):
...
+        # todo: @maethor: why searching with ack:false ? an element is important whether it is ack or not...
+        search = "bi:>%d " % self.min_business_impact + 'ack:false type:%s' % type
+        return self.search_hosts_and_services(search, user=user, sorter=sorter)
 
-    def get_impacts(self, user, search='is:impact bi:>=0 type:all', sorter=worse_first):
+    def get_impacts(self, user, search='is:impact type:all', sorter=worse_first):
         if "is:impact" not in search:
             search = "is:impact " + search
+        if "bi:" not in search:
+            search = "bi:>=%d " % self.min_business_impact + search
```
Comment (maethor) [inline module/datamanager.py]: Please put parenthesis around the % operation, for readability.

File: module/datamanager.py [hunk 2]
```diff
@@ -1000,21 +1024,42 @@ def get_overall_it_state(self, user):
...
         return self.search_hosts_and_services(search, user=user, get_impacts=True, sorter=sorter)
 
-    def get_problems(self, user, search='isnot:UP isnot:OK isnot:PENDING bi:>=0 type:all',
-                     get_acknowledged=False, get_downtimed=False, sorter=worse_first):
+    def get_problems(self, user, search='isnot:UP isnot:OK isnot:PENDING isnot:UNKNOWN isnot:ACK isnot:DOWNTIME '
+                                        'is:HARD type:all', get_acknowledged=False, get_downtimed=False,
+                     sorter=worse_first):
```
Comment (maethor) [inline module/datamanager.py:1077]: I disagree. UNKNOWN IS a problem. You can get an UNKNOWN when a check plugin doesn't work, if you filter on this you will never see that the plugin is not working (and the service may not be OK). Worse, if you don't configure thresholds properly, you can get an UNKNOWN on `check_http` when the site timeout. What does it mean ? Is it not a problem ?

I cannot merge that. Please don't change this query.

Comment (maethor) [reply module/datamanager.py:1077]: Please do not remove the bi filter. As discussed by email, you need to differentiate "min_business_impact" for the top badges and for the default problems query.

In my case, I want to keep a look on the "low importance" problems, like pending upgrades or problems on staging environments. But I don't want this problems to be counted in the top badges, because they do not need immediate attention or fixing.

File: module/datamanager.py [hunk 3]
```diff
@@ -1000,21 +1024,42 @@ def get_overall_it_state(self, user):
...
+        if "isnot:UNKNOWN" not in search:
+            search = "isnot:UNKNOWN " + search
+        if "isnot:ACK" not in search:
+            search = "isnot:ACK " + search
+        if "isnot:DOWNTIME" not in search:
+            search = "isnot:DOWNTIME " + search
+        if "is:HARD" not in search:
+            search = "is:HARD" + search
+        if "bi:" not in search:
+            search = "bi:>=%d " % self.min_business_impact + search
```
Comment (maethor) [inline module/datamanager.py]: NO! Why are you doing all this!?! No!

File: module/views/header_element.tpl [hunk 4]
```diff
@@ -37,7 +37,7 @@
        <table class="table table-invisible table-condensed">
          <tbody>
             <tr>
-               %for state in "up", "unreachable", "down", "pending", "unknown", "ack", "downtime":
```
Comment (maethor) [inline module/views/header_element.tpl:40]: Why did you removed pending?

File: module/datamanager.py [hunk 5]
```diff
@@ -121,17 +122,15 @@ def get_percentage_hosts_state(self, user=None, problem=False):
...
-        all_hosts = self.get_hosts(user=user)
-        if not all_hosts:
+        h = self.get_hosts_synthesis(None, user)
+        if not h['nb_elts']:
             return 0
 
-        problems = [h for h in all_hosts if h.is_problem and not h.is_impact]
+        known_problems = h['nb_problems']
+        if not problem:
+            known_problems = h['nb_elts'] - h['nb_problems']
```
Comment (maethor) [inline module/datamanager.py]: `not problem` is default in the function definition. So please begin by `if not problem: … else: …`

**Comment by maethor** [comment]
I'm sorry I cannot accept this PR. I leave you my comments in the review.

File: module/datamanager.py (hunk 1 above)
Comment (mohierf) [reply module/datamanager.py]: Ok

File: module/datamanager.py (hunk 2 above)
Comment (mohierf) [reply module/datamanager.py:1077]: It was a suggestion.

1. for UNKNOWN, it is a matter of view ... I can accept your analysis. I proposed to exclude UNKNOWN because Nagios does ... but I agree with you. I will revert this

1. for the BI filter I also agree. I propose another approachh ... each contact in the configuration has a `min_business_impact`  in its attributes and we should use this as a default filte for the problems view. The default value is 0 to view all the problems but if you set 2 for the user of the webui you will have a de-facto filter. It would make sense ...

File: module/datamanager.py (hunk 3 above)
Comment (mohierf) [reply module/datamanager.py]: The*get_problems* function is only used in the Wall view and it needs to be consistent with all the UI ... why forcing only 3 criteria whereas more are necessary to qualify a problem ...

File: module/views/header_element.tpl (hunk 4 above)
Comment (mohierf) [reply module/views/header_element.tpl:40]: Pending is only a transient state that is not not of a great interest in the pop over ...  this state only existswhen a check has not yet been launched by Shinken. If a host/service has this state it means that your Shinken stopped just after the WebUI staterd :/

**Comment by mohierf** [comment]
No problem that you did not accepted, it is the interest of the review 😉 and I appreciate your comments

File: module/datamanager.py (hunk 5 above)
Comment (mohierf) [reply module/datamanager.py]: Ok. I will change the test

File: module/views/header_element.tpl (hunk 4 above)
Comment (maethor) [reply module/views/header_element.tpl:40]: OK.

File: module/datamanager.py (hunk 3 above)
Comment (maethor) [reply module/datamanager.py]: OK.

File: module/datamanager.py (hunk 2 above)
Comment (maethor) [reply module/datamanager.py:1077]: About point 2. it makes sense. But I think you still need to have two variables, one for problems view and one for the badge.

Comment (mohierf) [reply module/datamanager.py:1077]: Hmmm... I will also check if the datamanger considers the current user `min_bi`. And then we will have this order in the filtering:
 - user `min_bi` filtering first
 - main bi filtering for the whole UI (including problems view)
 - badge bi filtering for the badges

**Comment by mohierf** [comment]
Concerning the BI filters, the user `min_business_impact` filters the default UI ...

For the same configuration, the *admin* user is default (0)
---
image

whereas the *guest* user is important (3)
---

image

**Comment by mohierf** [comment]
The badge filtering when set to Very important (4)
image

Or set to normal (2):
image

**Comment by mohierf** [comment]
Hosts / services global status remain like all the navbar icon, but the badge (if present...) is colored warning or critical:
image


Looks better 😉

**Comment by mohierf** [comment]
I rebased on the most recent develop branch. Should be merged now 😉

File: module/datamanager.py [hunk 6]
```diff
@@ -1000,21 +1053,41 @@ def get_overall_it_state(self, user):
...
+                                        'is:HARD type:all', get_acknowledged=False, get_downtimed=False,
+                     sorter=worse_first):
+        """What are problems:
+        - hosts/services (type:all) that are not OK/UP in a HARD state (isnot:hUP isnot:hOK)
+        - and that are not PENDING
+        - and which business impact is greater than or equal the minimum business impact (defaults 0)
+        """
         if "isnot:UP" not in search:
-            search = "isnot:UP " + search
+            search = search + " isnot:UP"
```
Comment (maethor) [inline module/datamanager.py]: Why did you change this? It is way better to filter first on "isnot:UP" because it is one of the fastest and more discriminent filter.

File: module/module.py [hunk 7]
```diff
@@ -341,6 +344,18 @@ def __init__(self, modconf):
         logger.info("[WebUI] parameter timezone: %s", self.timezone)
 
         # Visual alerting thresholds
+        # --------------------------
+        # All the hosts and services that are in a HARD non OK/UP state are considered as problems if their
+        # business_impact is greater than or equal this value
+        self.min_business_impact = int(getattr(modconf, 'min_business_impact', '0'))
```
Comment (maethor) [inline module/module.py]: Should be named `problems_min_business_impact` or maybe `problems_business_impact`.

File: module/module.py [hunk 8]
```diff
@@ -341,6 +344,18 @@ def __init__(self, modconf):
...
+        # badge_business_impact is used to filter the alerting badges in the header bar (default is 2)
+        self.badge_business_impact = int(getattr(modconf, 'badge_business_impact', '2'))
```
Comment (maethor) [inline module/module.py]: Should be named `important_problems_min_business_impact` or maybe `important_problems_business_impact`.

These variables should be named with "user meaning" in mind, not technical stuff or visual element like "badge". Because the point is to determinate which alert is a problem, and which is an important problem. Once you have this info, you use it anyway you need. Today it is for the badge in the header bar, but tomorrow it may be somewhere else.

File: module/module.py [hunk 9]
```diff
@@ -341,6 +344,18 @@ def __init__(self, modconf):
...
+        logger.info("[WebUI] minimum business impacts, all UI: %s, badges: %s",
+                    self.min_business_impact, self.badge_business_impact)
+
+        # Inner computation rules for the problems
+        self.inner_problems_count = int(getattr(modconf, 'inner_problems_count', '1'))
```
Comment (maethor) [inline module/module.py]: Why do you leave this option to user? Can't we just abandon the old way and adopt the new one if it is better?

If you need to keep this option, please change it to `enable_inner_problems_count`. Because `inner_problems_count` evokes a counter, not a boolean.

File: module/module.py (hunk 7 above)
Comment (maethor) [reply module/module.py]: Should default to 1 for consistent behavior, and because 0 officially means "no-importance" in Shinken docs.

File: module/plugins/problems/problems.py [hunk 10]
```diff
@@ -40,6 +40,10 @@ def get_page():
 
 def get_all():
     user = app.bottle.request.environ['USER']
+
+    # Update the default filter according to the logged-in user minimum business impact
+    default_filtering = DEFAULT_FILTER + " bi:>=%d" % (user.min_business_impact)
+
```
Comment (maethor) [inline module/plugins/problems/problems.py:46]: May not work if `user.min_business_impact` is lesser than `app.min_business_impact`. I didn't check.

Comment (maethor) [reply module/plugins/problems/problems.py:46]: Please be consistent in variable names. You can name this variable `default_filter` instead of `default_filtering`. By using lowercase instead of caps we will understand it is a local version of the variable.

File: module/views/_filters.tpl [hunk 11]
```diff
@@ -24,7 +24,8 @@
               <li role="presentation"><a role="menuitem" href="/all?search=type:host&title=All hosts">All hosts</a></li>
               <li role="presentation"><a role="menuitem" href="/all?search=type:service&title=All services">All services</a></li>
               <li role="presentation" class="divider"></li>
-              <li role="presentation"><a role="menuitem" href="/all?search=isnot:0 isnot:ack isnot:downtime&title=New problems">New problems</a></li>
+              <li role="presentation"><a role="menuitem" href="/all?search=isnot:0 is:hard isnot:ack isnot:downtime&title=New problems">New problems</a></li>
+              <li role="presentation"><a role="menuitem" href="/all?search=isnot:0 is:soft isnot:ack isnot:downtime&title=Potential problems">Potential problems</a></li>
```
Comment (maethor) [inline module/views/_filters.tpl:28]: So in "New problems" you are displaying PENDING, but not SOFT? Besides this, "New problems" is the same search than /problems. I don't see the point of keeping it.

File: module/views/header_element.tpl [hunk 12]
```diff
@@ -33,11 +33,14 @@
      <!-- Right part ... -->
      %s = app.datamgr.get_services_synthesis(user=user)
      %h = app.datamgr.get_hosts_synthesis(user=user)
+     %# Override the default problems count - badge minimum business impact
+     %h['nb_problems'] = len(app.datamgr.get_problems(user,'type:host bi:>%d' % app.badge_business_impact))
+     %s['nb_problems'] = len(app.datamgr.get_problems(user,'type:service bi:>%d' % app.badge_business_impact))
```
Comment (maethor) [inline module/views/header_element.tpl]: `user.min_business_impact` should be used if it is greater than `badge_business_impact`.

`max(user.min_business_impact, app.badge_business_impact)` may do the trick.

Better, I think the datamanager should store the `badge_business_impact` (`important_problems_business_impact`) variable, and use directly the `user.min_business_impact` variable. Here you should just need to call `app.datamgr.get_important_problems(user, 'type:host')`.

Even better, why do you need to add this in synthesis. Why `get_hosts_synthesis` and `get_services_synthesis` did not already compute the right count?

**Comment by maethor** [comment]
@mohierf OK. I review everything one more time. I leave some comments. Once this is done I will merge this PR.

File: module/datamanager.py (hunk 6 above)
Comment (mohierf) [reply module/datamanager.py]: +1 for you! I did a bad copy/paste ... I fix this and I will rebase/squash the commits for you merge this

**Review by maethor — CHANGES_REQUESTED** [review]

File: module/module.py (hunk 7 above)
Comment (mohierf) [reply module/module.py]: Ok for renaming and default value

File: module/module.py (hunk 9 above)
Comment (mohierf) [reply module/module.py]: I keep this option configurable because it may be interesting to have a view on the real computation done by the framewrok. 
I will rename as `disable_inner_problems_computation`

File: module/plugins/problems/problems.py (hunk 10 above)
Comment (mohierf) [reply module/plugins/problems/problems.py:46]: Note that `default_filtering`is only a local variable ... and this is only used to set a bollean value for the vie template. No search is executed with this criterai ... only set True/False for the empty search message in the problems view.

No modification in the code for this comment

File: module/module.py (hunk 8 above)
Comment (mohierf) [reply module/module.py]: Ok. I agree wit this. I will rename as `important_problems_business_impact`

File: module/views/_filters.tpl (hunk 11 above)
Comment (mohierf) [reply module/views/_filters.tpl:28]: Yes.
**Nxew problems** are problems not yet managed (HARD not OK/UP which are not yet ack/downtime)
**Potential problems** are the same but not yes confirmed (HARD) ...

File: module/views/header_element.tpl (hunk 12 above)
Comment (mohierf) [reply module/views/header_element.tpl]: I agree with your comment ... obviously, the best is refactoring of the synthesis functions to integrate all this stuff ... using a simple boolean parameter for important or not proiblems 😉 

 Ipropose to log an issue for this as a future enhancement
